devtools/zenodo/zenodo_data_release.py

```diff
@@ -0,0 +1,295 @@
+        return EmptyDraft(record_id=new_rec_id, zenodo_client=self.zenodo_client)
+
+
+class EmptyDraft(State):
+    """We can only sync the directory once we've gotten an empty draft."""
+
+    def sync_directory(self, source_dir: SourceData) -> "CompleteDraft":
+        """Read data from source_dir and upload it."""
+        logger.info(f"Syncing files from {source_dir} to draft {self.record_id}...")
+        for name, blob in source_dir.iter_files():
+            self.zenodo_client.create_deposition_file(self.record_id, name, blob)
+
+        return CompleteDraft(record_id=self.record_id, zenodo_client=self.zenodo_client)
+
+
+class CompleteDraft(State):
+    """Now that we've uploaded all the data, we can publish."""
+
+    def publish(self) -> None:
+        """Publish the draft.
+
+        We need to make sure there is complete metadata, including a publication date.
+
+        To do this, we:
+
+        1. use the *legacy* API to get the concept record ID associated with the draft
+        2. use the *new* API to get the latest record associated with the concept
+        3. use the *legacy* API to get the metadata from the latest record
+        4. use the *legacy* API to update the draft's metadata
+
+        Since we are using the legacy API to publish, we need the legacy
+        metadata format. But the legacy concept DOI -> published record mapping
+        is broken, so we have to take a detour through the new API.
+        """
+        deposition_info = self.zenodo_client.get_deposition(self.record_id).json()
+        concept_rec_id = deposition_info["conceptrecid"]
+        concept_info = self.zenodo_client.get_record(concept_rec_id).json()
+        latest_published_deposition = self.zenodo_client.get_deposition(
+            concept_info["id"]
+        ).json()
+        base_metadata = {
+            k: v
+            for k, v in latest_published_deposition["metadata"].items()
+            if k not in {"doi", "prereserve_doi", "publication_date"}
+        }
+        logger.info(
+            f"Using metadata from {latest_published_deposition['id']} to publish {self.record_id}..."
+        )
+        pub_date = {"publication_date": datetime.date.today().isoformat()}
+        self.zenodo_client.update_deposition_metadata(
+            self.record_id, metadata=base_metadata | pub_date
+        )
+        self.zenodo_client.publish_deposition(self.record_id)
+
+    def get_html_url(self):
+        """A URL for viewing this draft."""
+        return self.zenodo_client.get_deposition(self.record_id).json()["links"]["html"]
+
+
+@click.command()
+@click.option(
+    "--env",
+    type=click.Choice([SANDBOX, PRODUCTION], case_sensitive=False),
+    default=SANDBOX,
+    help="Use the Zenodo sandbox rather than the production server.",
+)
+@click.option(
+    "--concept-rec-id",
+    type=int,
+    required=True,
+    help="The concept record ID we'd like to sync to.",
+)
+@click.option(
+    "--source-dir",
+    type=Path,
```

Comment:
Could be worth using the `click.Path` type here, which makes it easy to verify that this path exists, is a directory, and is readable.

Reply:
Ah, I guess if we want this to work with `gs://` or `s3://` paths then we'd need to use an `fsspec` path here. It looks like they're done some work to try and provide a `pathlib.Path` like interface for remote filesystems as well. But it's not super mature. https://github.com/fsspec/universal_pathlib

Reply:
We can worry about `fsspec` when we get there... for now `click.Path` seems totally fine.